com.playeveryware.eos/Runtime/Core/AuthenticationListener.cs

```diff
@@ -42,7 +51,7 @@ public class AuthenticationListener: IAuthInterfaceEventListener, IConnectInterf
         /// True if the authentication state has changed to authenticated, False
         /// otherwise.
         /// </param>
-        public delegate void AuthenticationChangedEventHandler(bool authenticated);
+        public delegate void AuthenticationChangedEventHandler(bool authenticated, AuthenticationLevelChangeType changeType);
```

Comment:
Since the introduction and increased use of the `event` concept within our project, I've had the opportunity to learn more about common practices for implementing them effectively. If we're considering expanding or refactoring these implementations, I'd like to propose a refined approach based on these learnings:

## 1. Implement `AuthenticationChangedEventArgs` Class:

Within the `AuthenticationListener`, we can introduce a class called `AuthenticationChangedEventArgs`:

```csharp
public class AuthenticationChangedEventArgs : EventArgs
{
    public AuthenticationChangeOrigin Origin { get; }
    public bool IsAuthenticated { get; }

    public AuthenticationChangedEventArgs(bool isAuthenticated, AuthenticationChangeOrigin origin)
    {
        IsAuthenticated = isAuthenticated;
        Origin = origin;
    }
}
```

## 2. Use the Built-in `EventHandler<TEventArgs>` Delegate:**

By utilizing the built-in `EventHandler<TEventArgs>` delegate, we can simplify our event definition:

```csharp
public event EventHandler<AuthenticationChangedEventArgs> AuthenticationChanged;
```

**3. Update Event Invocation:**

When invoking the `AuthenticationChanged` event, we can adjust the invocation to:

```csharp
AuthenticationChanged?.Invoke(this, new AuthenticationChangedEventArgs(attemptedState, changeType));
```

> [!NOTE]
> The inclusion of the `sender` parameter (typed as `object`) in the `EventHandler<TEventArgs>` delegate allows us to track **who** is raising the event. This is particularly beneficial for debugging and provides flexibility if we need to access the sender down the call chain.

---

**Explanation and Benefits:**

- **Alignment with Standard Practices:** This approach adheres to established .NET conventions for event handling, which can improve code readability and maintainability.
  
- **Simplification:** Using `EventHandler<TEventArgs>` eliminates the need to define custom delegate types, streamlining our codebase.

- **Enhanced Debugging and Extensibility:** Passing the `sender` parameter and utilizing `EventArgs` derivatives provides context and allows for easier extension of event data in the future.

**Final Thoughts:**

I drafted the above without the aid of Visual Studio, so there might be minor implementation details to adjust.

Reply:
This was discussed within the team - and is something that we should consider in the future - but it was deemed that now was not the right time to consider these changes. Consequently, I am marking this conversation as resolved.